### ISSUE TITLE
Cannot create SPTPlaylistSnapshot from SPTPartialPlaylist

### ISSUE THREAD

**Issue by Walt-F4** [spotify/ios-sdk#179]
Hello all,

I'm having trouble creating a ```SPTPlaylistSnapshot``` object. Honestly I'm not sure I'm going at it correctly, but since the documentation points out that the snapshot inherits from ```SPTPartialPlaylist```, I assumed that the method to create a snapshot was ```SPTRequest```'s ```requestItemFromPartialObject:withSession:callback:```. Is this not the case?

I've also tried passing the ```SPTPartialPlaylist``` URI property to the ```requestItemAtURI``` method, as well as ```SPTPlaylistSnapshot```'s ```playlistWithURI:session:callback:``` using the partial playlist's URI, but neither of those worked either.

All methods return the following error:

    Error Domain=NSURLErrorDomain Code=-1012 "The operation couldn’t be completed. (NSURLErrorDomain error -1012.)" 
    UserInfo=0x19402fc0 {NSErrorFailingURLKey=https://api.spotify.com/v1/users/lordemusic/playlists/43Yim9aRqJJ5Rmq5ag8XRJ,
    NSErrorFailingURLStringKey=https://api.spotify.com/v1/users/lordemusic/playlists/43Yim9aRqJJ5Rmq5ag8XRJ,
    NSUnderlyingError=0x1908e9d0 "The operation couldn’t be completed. (kCFErrorDomainCFNetwork error -1012.)"}

I know that I'm passing a valid spotify:* URI so I'm not sure if I'm doing something wrong or if this is a bug of some kind. Any help is appreciated, thanks very much! 

**Edit:** I'm finding that ```NSURLErrorDomain Code=-1012``` is an authentication error, which made me think my scopes were wrong. They were, and I've updated them to include the ```SPTAuthPlaylistModifyPublicScope``` but I'm still getting the same error. Thoughts?

**Comment by possan** [comment]
You can use `requestItemAtURI` in `SPTRequest` to request a playlist, that will return a `SPTPlaylistSnapshot`

Modifying the example code like works:
```
[SPTRequest requestItemAtURI:[NSURL URLWithString:@"spotify:user:lordemusic:playlist:43Yim9aRqJJ5Rmq5ag8XRJ"]
                         withSession:session
                            callback:^(NSError *error, id object) {
            if (error != nil) {
                NSLog(@"*** Album lookup got error %@", error);
                return;
            }
            NSLog(@"playlist: %@", object);
            [self.player playTrackProvider:(id <SPTTrackProvider>)object callback:nil];
        }];
```

**Comment by Walt-F4** [comment]
Hey posson, thanks for the response. Good to know which request I should be using!

Unfortunately I'm still getting the same error with the code you posted. Error -1012. Do you think that is indeed a scope problem? I'm getting my session with these scopes: `SPTAuthStreamingScope, SPTAuthPlaylistModifyPublicScope, SPTAuthUserReadEmailScope, SPTAuthUserLibraryModifyScope`

I'm not sure why else I'd be getting an authorization error. Thanks for your help!

**Comment by gbrickley** [comment]
Any resolution on this?  I am trying to request a playlistSnapshot from a partialPlaylist as well with no luck.  The strange thing is that sometime it works and sometimes it doesn't (even for the same partialPlaylist in the same session).

I'm getting an error: -[NSNull longValue]: unrecognized selector

In the debugger it shows -[SPTPlaylistSnapshot initWithDecodedJSONObject:error:] as the problem.  I have confirmed that the partial object is in fact an SPTPartialPlaylist, my session is valid, and that I have correct scopes.  

Here's the code I'm using:

    // Create a full playlist from the partial playlist
    [SPTRequest requestItemFromPartialObject:partialPlaylist withSession:appDelegate.spotifySession callback:^(NSError *error, SPTPlaylistSnapshot *playlistToPass) {

        if (error != nil) {
            
            NSLog(@"Error: %@", error.userInfo[NSLocalizedDescriptionKey]);
            return;
            
        }

        self.tracksToPass = playlistToPass.firstTrackPage.items;
        self.playlistNameToPass = playlistToPass.name;
        
        [self performSegueWithIdentifier: @"playlistTracksSegue" sender:self];
   
    }];



Please help!

**Comment by Walt-F4** [comment]
Hey gbrickley, sorry you're having a similar issue. Did you try using the request that possan suggested above? Request at URI, rather than with a partial object?

**Comment by gbrickley** [comment]
Request using the URI is working perfectly, thanks!

**Comment by Walt-F4** [comment]
Glad to hear that! I guess I'm the only one having issues then, haha. What scopes are you using? I'm still getting Error -1012 no matter which scopes I use (and naturally I'm re-authorizing my session every time I change the scopes.) 

Would you mind posting the code that worked for you? Don't think you did anything different but it's worth a shot.

**Comment by gbrickley** [comment]
Here's the code I'm using now.  I'm grabbing all the users playlists and then one by one upgrading them to full playlists and replacing the partial playlist with the full playlist.  Already having a full playlist object seems to make a big difference in speed when the user clicks on a playlist to view the tracks (as opposed to creating the full playlist item on the click).

-(void)upgradeToFullPlaylists {

    if ([self.spotifyPartialPlaylistListItems count] == 0) {
        
        NSLog(@"Attempted to load all playlists!");
        return;
        
    }
    
    SPTPartialPlaylist *playlistToUpgrade = self.spotifyPartialPlaylistListItems[0];
    [self.spotifyPartialPlaylistListItems removeObjectAtIndex:0];
    
    [SPTRequest requestItemAtURI:playlistToUpgrade.uri withSession:self.spotifySession callback:^(NSError *error, SPTPlaylistSnapshot *playlistSnapshot) {
        
        if (error != nil) {
            
            NSLog(@"Error upgrading partial playlist! %@", error.userInfo[NSLocalizedDescriptionKey]);
            
        }
        
        else {
            
            NSLog(@"Successfully upgraded to full playlist: %@", playlistSnapshot);

            for (SPTPartialPlaylist *playlist in self.spotifyPlaylists) {
                
            if ([playlist.name isEqualToString:playlistSnapshot.name] && ![playlist isKindOfClass:[SPTPlaylistSnapshot class]]) {
                
                NSUInteger indexToReplace = [self.spotifyPlaylists indexOfObject:playlist];
                
                [self.spotifyPlaylists removeObjectAtIndex:indexToReplace];
                [self.spotifyPlaylists insertObject:playlistSnapshot atIndex:indexToReplace];

                break;

                }
                
            }
            
        }
        
        // Check if we've upgraded all the playlists
        self.spotifyPlaylistsLeftToUpgrade = self.spotifyPlaylistsLeftToUpgrade - 1;
        NSLog(@"%ld playlists left to upgrade.", (long)self.spotifyPlaylistsLeftToUpgrade);
            
        if (self.spotifyPlaylistsLeftToUpgrade <= 0 ) {
                
            NSLog(@"All playlists have been upgraded!");
            NSLog(@"Playlists array: %@", self.spotifyPlaylists);
                
        }
        
        else {
            
            [self upgradeToFullPlaylists];
                
        }
        
    }];
 
}

**Comment by Walt-F4** [comment]
Hmm, ok, so you're getting it in pretty much the exact same way. Really don't know why this isn't working for me. I'm logging out the URI so I know it's valid, but no matter what scopes I set I'm getting the same error. I'll post the code in case I'm missing something totally obvious here. 


```objectivec
SPTPartialPlaylist *playlist = self.playlists.items[indexPath.row];
NSLog(@"playlist uri: %@", playlist.uri);

dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        
    [SPTRequest requestItemAtURI:playlist.uri withSession:self.appDelegate.session callback:^(NSError *error, SPTPlaylistSnapshot *playlistSnapshot) {
                                
        if ( error != nil ) {
            NSLog(@"Error getting playlist snapshot: %@", error);
            [self showError];
            return;
        }

        // ... do stuff with snapshot object
    }];
});
```

**Comment by gbrickley** [comment]
Have you tried without the dispach_async part?  I really don't know much about that but the SPTRequest already handles running on a background queue so maybe its an issue caused by changing queues.  My only guess...

**Comment by Walt-F4** [comment]
Oh does it? Well that's handy. Still getting the error though :/

**Comment by iKenndac** [comment]
Yes, the library deals with threading for you, hence all the block callbacks. There's zero need to separately dispatch calls to it. This is *especially* true with playback, and dispatching there will likely cause problems.

**Comment by Walt-F4** [comment]
Awesome, well that's great to know! Thanks guys. So, iKenndac, do you have any idea why I'd be getting an Error -1012, then? Any idea why this error crops up in other places, maybe? If it helps, I get the error almost immediately after requesting the playlist snapshot. No trouble retrieving the playlists but I get the error immediately upon requesting a snapshot with one of the retrieved URIs.

**Comment by iKenndac** [comment]
-1012 is `NSURLErrorUserCancelledAuthentication`, which can be caused by: Not having the correct scopes (try adding every single playlist related one), SSL problems (is your system's time correct?), etc.

**Comment by iKenndac** [comment]
…also, when you change your scopes, you need to completely throw away your saved tokens and authenticate again from scratch.

**Comment by Walt-F4** [comment]
Ok, so it almost certainly is a scope problem then. I don't think it's SSL because this is the only place I'm getting this error. (Is that a safe assumption?)

I've tried reauthorizing with the new scopes, but nothing changes. I open the login URL, create a new session, all that -- is there some way to "flush" the old tokens that I'm missing? I know that sounds nonsensical but I get the impression it's using the old scopes: as a test, I removed the streaming scope but was still able to stream music. Why would it be using old scopes? 

Thanks again for your help!

**Comment by Walt-F4** [comment]
Hey guys, just bumping this again because it's more or less the only substantial bug still in my app :smile: 

Like I said, I've added all playlist-related scopes and reauthorized but I'm still getting error -1012. I'm not sure I'm successfully flushing old scopes -- when I removed the streaming scope and logged in again, I could still stream. Any idea what I could be doing wrong / forgetting to do in "flushing" the old scopes? 

Thanks again!

**Comment by iKenndac** [comment]
Random guess — are you using a swap service that caches stuff? Maybe clearing out that database would help. I'll go point our web guys to this thread, see if they can help.

**Comment by possan** [comment]
There currently is no way to flush scopes that you already allowed once, so if you at one point added streaming scope for your user you will always get that (currently, this will probably change in the future) - the only way to flush scopes is to set up a new client id and test with that.

**Comment by Walt-F4** [comment]
Thanks for the speedy response. Daniel, that was an excellent tip about the database -- our swap script was in fact still communicating with a database. Unfortunately we removed database interaction from our scripts and tried again and it didn't address the problem.

The real kicker, though, was trying again with a whole new account (as you suggested, possan). I authorized an as-yet unused Spotify Premium account, pulled their playlists successfully, and then received the same -1012 error when attempting to create a playlist snapshot. 

I even created a new application, passed in the new client id and redirect url to the swap and the app code, then tried again, but no dice. Could stream and view playlists but not create a playlist snapshot.

At this point I'm at a bit of a loss. Is there any way of introspecting the current permissions? All playlist modification scopes showed up in the "This app will be able to..." list when I authorized the new user, so I know the scopes were at least passed correctly, but it'd be ideal to log out exactly what was allowed AFTER login. I'm starting to suspect that it's not a scope issue after all, rather some other authentication related error, but still, the scope seems like a likely culprit and it's really all I would know how to diagnose. 

For posterity, the code I'm using to set up the scopes, as well as the playlist snapshot request again:

```objc
// Open auth URL
SPTAuth *auth = [SPTAuth defaultInstance];
NSURL *loginURL = [auth loginURLForClientId:kClientId 
                       declaredRedirectURL:[NSURL URLWithString:kCallbackURL]
                                       scopes:@[SPTAuthStreamingScope, SPTAuthPlaylistReadPrivateScope, SPTAuthPlaylistModifyPublicScope, SPTAuthPlaylistModifyPrivateScope, SPTAuthUserLibraryReadScope, SPTAuthUserReadEmailScope]];
[[UIApplication sharedApplication] performSelector:@selector(openURL:) withObject:loginURL];

...

// Get playlist snapshot
SPTPartialPlaylist *playlist = self.playlists.items[indexPath.row];
NSLog(@"playlist uri: %@", playlist.uri); // returns a valid playlist uri

[SPTRequest requestItemAtURI:playlist.uri
              withSession:session
                    callback:^(NSError *error, SPTPlaylistSnapshot *playlistSnapshot) {

	 if ( error != nil ) {
		 NSLog(@"Error getting playlist snapshot: %@", error);
		 [self showError];
		 return;
	 }

	 NSLog(@"Success!");
}];
```

**Comment by gbrickley** [comment]
Out of curiosity, are you able to successfully request other objects from their URI like tracks or artists?  Have you double checked that your session is valid?

**Comment by Walt-F4** [comment]
![](http://i.imgur.com/mpma0eu.gif)

Ok, so, um... wow, I am a fool. gbrickley, I could kiss you.

I checked that I was getting a valid session, yes, and that the session had the correct scopes, but I never checked if I was *actually passing that valid session to the request.* Spoiler alert: I was not. There was an elementary variable problem that I cannot believe I missed. It's working now. 

Thanks to all for being so helpful, regardless. I learned a lot about threading, scopes, token swap, and the outer boundaries of my own stupidity. Cheers.
